### PR TITLE
Simple partition pruning

### PR DESCRIPTION
**Pull request by zhexuany**
Simple means partition expression pruning only works on the column, not a function. 

If we have the following table schema:
```
CREATE TABLE `pt` (
  `id` int(11) DEFAULT NULL,
  `y` date DEFAULT NULL,
  index `idx_y`(`y`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin
PARTITION BY RANGE (id ) (
  PARTITION p0 VALUES LESS THAN (2),
  PARTITION p1 VALUES LESS THAN (6),
  PARTITION p2 VALUES LESS THAN (10)
)
```

`select * from pt where id > 3 and id < 6` will only access p1. If partition expression is `year(y)` not `id`, then partition pruning will not work.

### CODE REVIEW

File: tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    return filter.accept(this, null);
+  }
+
+  private static List<Expression> generatePartExprs(TiTableInfo tableInfo) {
+    // only support column ref for now. Fn Expr will be supported later.
+    TiPartitionInfo partInfo = tableInfo.getPartitionInfo();
+    List<Expression> partExprs = new ArrayList<>();
+    for (int i = 0; i < partInfo.getDefs().size(); i++) {
+      TiPartitionDef pDef = partInfo.getDefs().get(i);
+      Constant cst = Constant.create(Longs.tryParse(pDef.getLessThan().get(0)));
```
Comment (ilovesoup) [inline tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java]: Again, why long? Even if current stage of TiDB implements long only, your implementation should not hard-code it like this.

File: tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+      if (i == 0) {
+        partExprs.add(ComparisonBinaryExpression.lessThan(partExpr, cst));
+      } else {
+        Constant previous =
+            Constant.create(Longs.tryParse(partInfo.getDefs().get(i - 1).getLessThan().get(0)));
+        Constant current =
+            Constant.create(Longs.tryParse(partInfo.getDefs().get(i).getLessThan().get(0)));
```
Comment (ilovesoup) [inline tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java]: Would you please explain why get(0)? If you are only to deal with range type partition only, please at least leave room for other type in your code. If this method is just to generate expr for range type, name it like generateRangePartExpr. Also guard on partition type at beginning.

File: tikv-client/src/main/java/com/pingcap/tikv/meta/TiColumnInfo.java [hunk 3]
```diff
@@ -130,7 +130,8 @@ public String getName() {
   }
 
   public boolean matchName(String name) {
-    return this.name.equalsIgnoreCase(name);
+    return this.name.equalsIgnoreCase(name)
+        || String.format("`%s`", this.name).equalsIgnoreCase(name);
```
Comment (ilovesoup) [inline tikv-client/src/main/java/com/pingcap/tikv/meta/TiColumnInfo.java]: When will a column be quoted?

File: tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+      if (expr instanceof Not || expr instanceof IsNull) {
+        continue;
+      }
+      filteredFilters.add(expr);
+    }
+    return filteredFilters;
+  }
+
+  @Override
+  @SuppressWarnings("Duplicates")
```
Comment (ilovesoup) [inline tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java]: Is it possible to reuse index range builder code?

File: tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    for (int i = 0; i < partExprs.size(); i++) {
+      Expression partExpr = partExprs.get(i);
+      RangeSet<Long> partRange = buildPartRange(partExpr);
+      partRange.removeAll(filterRange.complement());
+      if (!partRange.isEmpty()) {
+        // part range is empty indicates this partition can be pruned.
+        pDefs.add(partInfo.getDefs().get(i));
+      }
+    }
+    return new TiPartitionInfo(
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java]: Partition type info is also needed.

File: tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java (hunk 1 above)
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java]: For range partition, only int or function return int can be accepted as partition expression.

File: tikv-client/src/main/java/com/pingcap/tikv/meta/TiColumnInfo.java (hunk 3 above)
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/meta/TiColumnInfo.java]:
```
CREATE TABLE `pt` (
  `id` int(11) DEFAULT NULL,
  `y` date DEFAULT NULL,
  KEY `idx_y` (`y`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin
PARTITION BY RANGE ( `id` + 1 ) (
  PARTITION p0 VALUES LESS THAN (2),
  PARTITION p1 VALUES LESS THAN (4),
  PARTITION p2 VALUES LESS THAN (6)
)
```

If we have above table schema, `partExpr` in `PartitionInfo` is "`id`". This is just a hard code version. When parser gets merged, I will rewrite this part.

File: tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java (hunk 4 above)
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java]: It is possible. I will work on that.

File: tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java (hunk 2 above)
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java]: It is hardcode version and will be replaced by a parser. The parser can handle this well. I should comment on this. Maybe we should first review parser code?

File: tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+    // if query is select * from t, then filter will be null.
+    if (filter == null) return false;
+
+    // Ideally, part expr should be parsed by a parser
+    // but it requires a lot work, we chose a brute force way
+    // to workaround.
+    boolean canBePruned = true;
+    partExpr = ColumnRef.create(tblInfo.getPartitionInfo().getExpr(), tblInfo);
+    if (!tblInfo.getPartitionInfo().getExpr().contains("(")) {
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java]: It should rebase master once parser code gets merged into master. - -

File: tikv-client/src/test/java/com/pingcap/tikv/expression/visitor/CanBePrunedValidatorTest.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+
+public class CanBePrunedValidatorTest {
+
+  @Test
+  public void canBePrunedOrWithUnrelatedColCase() {
+    TiTableInfo tableInfo = createTable();
+    Expression partExpr = ColumnRef.create("c1", tableInfo);
+    Expression ts = ColumnRef.create("c3", tableInfo);
```
Comment (birdstorm) [inline tikv-client/src/test/java/com/pingcap/tikv/expression/visitor/CanBePrunedValidatorTest.java]: I think they are ColumnRef

File: tikv-client/src/test/java/com/pingcap/tikv/expression/visitor/CanBePrunedValidatorTest.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
+package com.pingcap.tikv.expression.visitor;
+
+import static com.pingcap.tikv.expression.visitor.ExpressionTypeCoercerTest.createTable;
```
Comment (birdstorm) [inline tikv-client/src/test/java/com/pingcap/tikv/expression/visitor/CanBePrunedValidatorTest.java]: Seems a bit strange to use createTable in another test... We are only using c1 and c3 here.

File: tikv-client/src/test/java/com/pingcap/tikv/expression/visitor/CanBePrunedValidatorTest.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+import com.pingcap.tikv.expression.visitor.CanBePrunedValidator.ContextPartExpr;
+import com.pingcap.tikv.meta.TiTableInfo;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class CanBePrunedValidatorTest {
+
+  @Test
+  public void canBePrunedOrWithUnrelatedColCase() {
+    TiTableInfo tableInfo = createTable();
```
Comment (birdstorm) [inline tikv-client/src/test/java/com/pingcap/tikv/expression/visitor/CanBePrunedValidatorTest.java]: do we need to add partition info in createTable?

File: tikv-client/src/test/java/com/pingcap/tikv/predicates/ScanAnalyzerTest.java [hunk 10]
```diff
@@ -93,7 +93,7 @@ public void buildTableScanKeyRangeTest() throws Exception {
     ScanAnalyzer scanAnalyzer = new ScanAnalyzer();
 
     List<Coprocessor.KeyRange> keyRanges =
-        scanAnalyzer.buildTableScanKeyRange(table, irs, table.getPartitionInfo());
+        scanAnalyzer.buildTableScanKeyRange(table, irs, table.getPartitionInfo().getDefs());
```
Comment (birdstorm) [inline tikv-client/src/test/java/com/pingcap/tikv/predicates/ScanAnalyzerTest.java]: any test for buildTableScanKeyRange when the table contains PartitionInfo?

File: tikv-client/src/test/java/com/pingcap/tikv/expression/visitor/CanBePrunedValidatorTest.java (hunk 9 above)
Comment (zhexuany) [reply tikv-client/src/test/java/com/pingcap/tikv/expression/visitor/CanBePrunedValidatorTest.java]: No. This is just to verify the correctness of canBePrunedValidator .

File: tikv-client/src/test/java/com/pingcap/tikv/predicates/ScanAnalyzerTest.java (hunk 10 above)
Comment (zhexuany) [reply tikv-client/src/test/java/com/pingcap/tikv/predicates/ScanAnalyzerTest.java]: We do not need to. We just need to verify the correctness of pruning logic. 

There is no significant difference between the partitioned table and the normal table. If we meet a partitioned table, we only need to replace the table id with partition id.

Comment (birdstorm) [reply tikv-client/src/test/java/com/pingcap/tikv/predicates/ScanAnalyzerTest.java]: Then you may comment this :)

**Pull request closed without merging**